src/trusted-device/Security/TwoFactor/Trusted/JwtTokenEncoder.php

```diff
@@ -16,51 +19,55 @@ class JwtTokenEncoder
     public const CLAIM_VERSION = 'vsn';
 
     /**
-     * @var Sha256
+     * @var Configuration
      */
-    private $signer;
-
-    /**
-     * @var string
-     */
-    private $applicationSecret;
+    private $configuration;
 
     public function __construct(string $applicationSecret)
     {
-        $this->signer = new Sha256();
-        $this->applicationSecret = $applicationSecret;
+        $this->configuration = Configuration::forSymmetricSigner(new Sha256(), InMemory::plainText($applicationSecret));
+        $this->configuration->setValidationConstraints(new SignedWith($this->configuration->signer(), $this->configuration->signingKey()));
     }
 
     public function generateToken(string $username, string $firewallName, int $version, \DateTimeInterface $validUntil): Token
     {
-        $builder = (new Builder())
-            ->setIssuedAt(time())
-            ->setExpiration($validUntil->getTimestamp())
-            ->set(self::CLAIM_USERNAME, $username)
-            ->set(self::CLAIM_FIREWALL, $firewallName)
-            ->set(self::CLAIM_VERSION, $version)
-            ->sign($this->signer, $this->applicationSecret);
+        if ($validUntil instanceof \DateTime) {
+            $validUntil = \DateTimeImmutable::createFromMutable($validUntil);
+        } elseif (!$validUntil instanceof \DateTimeImmutable) {
+            throw new \InvalidArgumentException(sprintf('The 4th argument to "%s" must be instance of either a "DateTime" or "DateTimeImmutable", "%s" given.', __METHOD__, \get_class($validUntil)));
+        }
 
-        return $builder->getToken();
+        $builder = $this->configuration->builder()
+            ->issuedAt(new \DateTimeImmutable())
+            ->expiresAt($validUntil)
+            ->withClaim(self::CLAIM_USERNAME, $username)
+            ->withClaim(self::CLAIM_FIREWALL, $firewallName)
+            ->withClaim(self::CLAIM_VERSION, $version);
+
+        return $builder->getToken($this->configuration->signer(), $this->configuration->signingKey());
     }
 
     public function decodeToken(string $token): ?Token
     {
         try {
-            $token = (new Parser())->parse($token);
-        } catch (\InvalidArgumentException $e) {
+            $token = $this->configuration->parser()->parse($token);
+        } catch (Exception $e) {
             return null; // Could not decode token
         }
 
+        if (!$token instanceof Plain) {
+            return null;
+        }
+
         try {
-            if (!$token->verify($this->signer, $this->applicationSecret)) {
+            if (!$this->configuration->validator()->validate($token, ...$this->configuration->validationConstraints())) {
                 return null;
             }
         } catch (\BadMethodCallException $e) {
```

Comment:
I believe this check can go away, the new implementation doesn't seem to throw that exception anymore.